enhancements/cloud-integration/infrastructure-external-platform-type.md

```diff
@@ -0,0 +1,513 @@
+Related PlatformSpec and PlatformStatus type structures will serve as a source of generic configuration information for
+the Openshift-specific operators.
+
+Having that special platform type will allow infrastructure partners to clearly designate when their OpenShift
+deployments contain components that replaces and/or supplement the core Red Hat components.
+
+### Affected components
+
+This section enumerates OpenShift's components and briefly elaborates the future plans around this proposal.
+During [initial implementation](#implementation-phases) we must ensure that all OpenShift's components treat "External" platform in the same way as
+"None" to ensure we do not disrupt the current openshift operation.
+
+In the future, we will need to change the behavior of OpenShift components to be able to receive supplemental provider-specific components from an infrastructure provider
+or, if a component manages something else (i.e. kubelet, kcm), adjust its behaviour (set `--cloud-provider=external` arg to kubelet for example).
+
+Specific component changes will be described in detail within separate enhancement documents on a per-component basis.
+
+#### Machine Config Operator
+
+Currently, MCO sets kubelet flags based on a [PlatformStatus and PlatformType](https://github.com/openshift/machine-config-operator/blob/7babc03f84088f49072c6f5b0d1769850d09a694/pkg/controller/template/render.go#L343).
+This flag is [crucial](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager) for Cloud Controller Manager engagement within the cluster.
+
+Initially, the new "External" platform should be treated similarly to PlatformType "None" by the MCO, and do not set up any cloud-specific flags for the kubelet.
+Then, down the road (during [phase 3](#implementation-phases)), it would be expected for the MCO to use the "External" platform type and its spec as a signal about the underlying platform and cloud controller manager presence and operate accordingly.
+For an explicit signal about the necessity to set `--cloud-provider=external` flag to the kubelet, we will use the `CloudControllerManager` field of the `ExternalPlatformSpec`, which is  
+described in the [API Extensions](#api-extensions) section down below.
+
+#### Kube Controller Manager Operator
+
+Same as the [MCO](#machine-config-operator) in regard to kubelet, Kube Controller Manager Operator manages KCM (kube-controller-manager) deployments.
+
+Historically Kube Controller Manager was home for cloud-specific control loops. This logic is also engaging by setting up proper
+flags on KCM executable like,
+```shell
+...
+--cloud-provider=azure
+--cloud-config=/etc/kubernetes/cloud.conf
+...
+```
+
+For engaging an external Cloud Controller Manager no `cloud-provider` flag should be set for the KCM executable.
+
+In the context of this EP, no particular changes will be needed in the operator itself.
```

Comment:
iiuc, the logic in our KCMO will currently "do the right thing", but i think @lobziik has more context here. from looking at the code, if there is an explicit "external" cloud provider, or if the field is empty, then the KCM will react the same way. so, as i understand, we don't need to change the KCMO since the platform will never be recognized, and if an external cloud provider is specified then the CCCMO will do something and the KCMO should still do nothing.

that said, hopefully @lobziik will correct what i have gotten wrong ;)

Reply:
Almost all logic around this lives in library-go, in KCMO it should be just revendored. Will rewrite this sections.